respokeSDK/src/main/java/com/digium/respokesdk/RespokeClient.java

```diff
@@ -1171,4 +1399,43 @@ public void run() {
         });
     }
 
+
+    /**
+     * Build a group message from a JSON object. The format of the JSON object would be the
+     * format that comes over the wire from Respoke when receiving a pubsub message. This same
+     * format is used when retrieving message history.
+     *
+     * @param source The source JSON object to build the RespokeGroupMessage from
+     * @return The built RespokeGroupMessage
+     * @throws JSONException
+     */
+    private RespokeGroupMessage buildGroupMessage(JSONObject source) throws JSONException {
+        if (source == null) {
+            throw new IllegalArgumentException("source cannot be null");
+        }
+
+        final JSONObject header = source.getJSONObject("header");
+        final String endpointID = header.getString("from");
+        final RespokeEndpoint endpoint = getEndpoint(endpointID, false);
+        final String groupID = header.getString("channel");
+        RespokeGroup group = getGroup(groupID);
+
+        if (group == null) {
```

Comment:
I think you need to do similar treatment for endpoint... i.e. if it doesn't exist, create it.